provider/pkg/provider/invoke_helm_template.go

```diff
@@ -245,6 +242,43 @@ func (c *chart) template() (string, error) {
 	installAction.ReleaseName = c.opts.ReleaseName
 	installAction.Version = c.opts.Version
 
+	if clientSet != nil {
+		// The following code to discover Kubernetes version and API versions comes
+		//  from the Helm project:
+		// https://github.com/helm/helm/blob/d7b4c38c42cb0b77f1bcebf9bb4ae7695a10da0b/pkg/action/action.go#L239
+
+		dc := clientSet.DiscoveryClientCached
+
+		dc.Invalidate()
+		kubeVersion, err := dc.ServerVersion()
+		if err != nil {
+			return "", fmt.Errorf("could not get server version from Kubernetes: %w", err)
+		}
+		// Client-Go emits an error when an API service is registered but unimplemented.
+		// Since the discovery client continues building the API object, it is correctly
+		// populated with all valid APIs.
+		// See https://github.com/kubernetes/kubernetes/issues/72051#issuecomment-521157642
+		apiVersions, err := action.GetVersionSet(dc)
+		if err != nil {
+			if !discovery.IsGroupDiscoveryFailedError(err) {
+				return "", fmt.Errorf("could not get apiVersions from Kubernetes: %w", err)
+			}
+		}
+
+		installAction.KubeVersion = &chartutil.KubeVersion{
+			Version: kubeVersion.GitVersion,
+			Major:   kubeVersion.Major,
+			Minor:   kubeVersion.Minor,
+		}
+
+		installAction.APIVersions = apiVersions
+	}
+
+	// Preserve backward compatibility
+	if len(c.opts.APIVersions) > 0 {
+		installAction.APIVersions = c.opts.APIVersions
+	}
```

Comment:
I'd check this first, and skip the discovery client calls to the apiserver if true.

Reply:
Hello @lblackstone, thanks for the review :wink:

I followed your suggestion and modified the code, but now I’m wondering if we really want to disable Kubernetes version discovery when APIVersions are provided. What do you think?

Regarding the tests, it seems they haven’t run, do I have to fix something on my side?

Reply:
I posted a message in Slack (https://pulumi-community.slack.com/archives/CRFURDVQB/p1661613221358179) to get some directions on how to develop on this provider, maybe you have the answers to my questions?

Reply:
> I followed your suggestion and modified the code, but now I’m wondering if we really want to disable Kubernetes version discovery when APIVersions are provided. What do you think?

Ah, I didn't notice that the `DefaultCapabilities` was setting the kube version before, and defaulted to v1.20. Maybe we should set that explicitly here? The problem is that the cluster may not be available yet when this code is running (i.e., if it's a preview in a stack where the cluster hasn't been created yet), so we need a fallback value set.

> Regarding the tests, it seems they haven’t run, do I have to fix something on my side?

For external contributors, a maintainer has to run the tests with the `/run-acceptance-tests` comment, and then the bot will respond with a link to the test run that you can check: https://github.com/pulumi/pulumi-kubernetes/pull/2155#issuecomment-1230941714

It looks like everything passed, so this should be good once we handle the default k8s version.

Reply:
> > I followed your suggestion and modified the code, but now I’m wondering if we really want to disable Kubernetes version discovery when APIVersions are provided. What do you think?
> 
> Ah, I didn't notice that the `DefaultCapabilities` was setting the kube version before, and defaulted to v1.20. Maybe we should set that explicitly here? The problem is that the cluster may not be available yet when this code is running (i.e., if it's a preview in a stack where the cluster hasn't been created yet), so we need a fallback value set.

The default KubeVersion seems to still be set here https://github.com/helm/helm/blob/dbc6d8e20fe1d58d50e6ed30f09a04a77e4c68db/pkg/action/install.go#L216.

My concern was about the fact that people having APIVersions manually set won’t benefit from the discovery of the KubeVersion. But it’s not a regression. We could also add the ability to manually set the KubeVersion (as for the APIVersions), but that’d be a fairly different PR :wink: 

> 
> > Regarding the tests, it seems they haven’t run, do I have to fix something on my side?
> 
> For external contributors, a maintainer has to run the tests with the `/run-acceptance-tests` comment, and then the bot will respond with a link to the test run that you can check: [#2155 (comment)](https://github.com/pulumi/pulumi-kubernetes/pull/2155#issuecomment-1230941714)
> 
> It looks like everything passed, so this should be good once we handle the default k8s version.

Ha OK, I had a notification with this link https://github.com/pulumi/pulumi-kubernetes/actions/runs/2939221543 which made me think the tests didn’t run, I didn’t see https://github.com/pulumi/pulumi-kubernetes/actions/runs/2952219436, so that’s OK :wink: